bootloader/lk.go

```diff
@@ -0,0 +1,202 @@
+	// at runtime environment is writen directly into dedicated partition
+	if dirs.GlobalRootDir == "/" {
+		return "/dev/disk/by-partlabel/"
+	} else {
+		return filepath.Join(dirs.GlobalRootDir, "/boot/lk/")
+	}
+}
+
+func (l *lk) ConfigFile() string {
+	return l.envFile()
+}
+
+func (l *lk) envFile() string {
+	// as for dir, we have two scenarios, image building and runtime
+	if dirs.GlobalRootDir == "/" {
+		// TO-DO: this should be eventually fetched from gadget.yaml
+		return filepath.Join(l.Dir(), "snapbootsel")
+	} else {
+		return filepath.Join(l.Dir(), "snapbootsel.bin")
+	}
+}
+
+func (l *lk) GetBootVars(names ...string) (map[string]string, error) {
+	out := make(map[string]string)
+
+	env := lkenv.NewEnv(l.envFile())
+	if err := env.Load(); err != nil {
+		return nil, err
+	}
+
+	for _, name := range names {
+		out[name] = env.Get(name)
+	}
+
+	return out, nil
+}
+
+func (l *lk) SetBootVars(values map[string]string) error {
+	env := lkenv.NewEnv(l.envFile())
+	if err := env.Load(); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	// update environment only if something change
+	dirty := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirty = true
+	}
+
+	if dirty {
+		return env.Save()
+	}
+
+	return nil
+}
+
+// for lk we need to flash boot image to free bootimg partition
+// first make sure there is free boot part to use
+// if this is image creation, we just extract file
+func (l *lk) ExtractKernelAssets(s *snap.Info, snapf snap.Container) error {
+	blobName := filepath.Base(s.MountFile())
+
+	logger.Debugf("ExtractKernelAssets (%s)\n", blobName)
+
+	env := lkenv.NewEnv(l.envFile())
+	if err := env.Load(); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	bootPartition, err := env.FindFreeBootPartition(blobName)
+	if err != nil {
+		return err
+	}
+
+	if dirs.GlobalRootDir == "/" {
+		logger.Debugf("ExtractKernelAssets handling run time usecase\n")
+		// this is live system, extracted bootimg needs to be flashed to
+		// free bootimg partition and env has be updated boot slop mapping
+		tmpdir, err := ioutil.TempDir("", "bootimg")
+		if err != nil {
+			return fmt.Errorf("Failed to create tmp directory %v", err)
+		}
+		defer os.RemoveAll(tmpdir)
+		if err := snapf.Unpack("boot.img", tmpdir); err != nil {
+			return fmt.Errorf("Failed to unpack boot.img %v", err)
+		}
+		// write boot.img to free boot partition
+		bootimgName := filepath.Join(tmpdir, "boot.img")
+		bif, err := os.Open(bootimgName)
+		if err != nil {
+			return fmt.Errorf("Failed to open unpacked boot.img %v", err)
+		}
+		defer bif.Close()
+		bpart := filepath.Join("/dev/disk/by-partlabel/",bootPartition)
+
+		bpf, err := os.OpenFile(bpart, os.O_WRONLY, 0660)
+		if err != nil {
+			return fmt.Errorf("Failed to open boot partition [%s] %v", bpart, err)
+		}
+		defer bpf.Close()
+
+		buf := make([]byte, 1024)
+		for {
```

Comment:
IDK, it feels like this should be a part of the upcoming gadget update. I assume boot.img is the actual image of the partition where `snapbootsel[.bin]` are in. Is this a workaround for deploying the actual thing to the device?

Reply:
No, none of this is really related to gadget assets. Only relation to gadget is gadget.yaml, where bootloader is defined (though not actually used by snapd to determine bootlloader type! :) ) And then added extra supported roles for partitions. This is to make it more dynamic rather than hardcore expected partition labels.
`snapbootsel` is direct equivalent of `uboot.env`, just stored as partition rather than file. So it's changed every time new core/kernel is installed/tried.
boot.img is then asset extracted from kernel snap. boot.img is actually kernel + initrd + dtbs packed into one file, and flashed into dedicated partition. So we need two partitions for boot.img which we will alter based on installed kernel snap revision.

So none is workaround, for future improvement I'd like bootloader implementation to more actively use info from gadget snap, to determine bootloader type for example.